system/src/commonMain/kotlin/com/splendo/kaluga/system/network/BaseNetworkManager.kt

```diff
@@ -0,0 +1,61 @@
+/*
+ Copyright 2020 Splendo Consulting B.V. The Netherlands
+
+    Licensed under the Apache License, Version 2.0 (the "License");
+    you may not use this file except in compliance with the License.
+    You may obtain a copy of the License at
+
+      http://www.apache.org/licenses/LICENSE-2.0
+
+    Unless required by applicable law or agreed to in writing, software
+    distributed under the License is distributed on an "AS IS" BASIS,
+    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+    See the License for the specific language governing permissions and
+    limitations under the License.
+
+ */
+
+package com.splendo.kaluga.system.network
+
+import kotlinx.coroutines.CoroutineScope
+import kotlinx.coroutines.launch
+
+abstract class BaseNetworkManager(private val networkStateRepo: NetworkStateRepo) : CoroutineScope by networkStateRepo {
+
+    internal abstract suspend fun startMonitoringNetwork()
+    internal abstract suspend fun stopMonitoringNetwork()
```

Comment:
These methods can be implemented in this base class, after all, they should use the state machine to make the changes.

Reply:
Also in `BaseLocationManager` are declared as `abstract` and `LocationManager` overrides them

Reply:
I can see the temptation to use it like that, and probably at some point myself choose this option in location manager (but before we had the more full featured state machine we do now). In particular for location on Android you have a broadcast listener etc that makes it hard to avoid this (though not impossible).

The problem is there is no guarantee these start/stop functions are called out of order since they are not part of the state machine. But by using the initialstate / laststate "trick" you implemented this is actually guarded pretty well (once it's deduplicated), so I guess it's acceptable.

That said, I don't see a reason why the staterepo itself can not create an expected manager (instead of the manager holding the state repo) when going hot (passing a new one into initial value, and this manager will start itself on construction, so it's guaranteed to be once only), and dispose it when going cold. This manager will not have to null its listeners and such stuff when going cold, since it will simply stop to exist. It fact it _might_ not even need all the atomics on iOS (since creation/access can be done from the same dispatcher).

The constructor would take a `(Network) -> Unit` and the `handleNetworkStateChanged` could move to the state repo itself. You can probably see that this will also decouple some things more.

If you want we can discuss it on Friday, or I can take a stab at it myself.

Reply:
acf6254 3cc127d cd7bf8c 9be9046